packages/eslint-plugin-testing-library-react/src/rules/prefer-expect-query-by.ts

```diff
@@ -0,0 +1,114 @@
+import {
+  AST_NODE_TYPES,
+  TSESTree,
+} from '@typescript-eslint/experimental-utils';
+import { createRule } from '../utils';
+
+interface ExtendedExpression extends TSESTree.CallExpression {
+  callee: TSESTree.MemberExpression | TSESTree.Identifier;
+}
+
+function isIdentifier(
+  node: ExtendedExpression['callee']
+): node is TSESTree.Identifier {
+  return node.type === AST_NODE_TYPES.Identifier;
+}
+
+function isMemberExpression(
+  node: ExtendedExpression['callee']
+): node is TSESTree.MemberExpression {
+  return node.type === AST_NODE_TYPES.MemberExpression;
+}
+
+function isUsingWrongQueries(node: TSESTree.Identifier) {
+  return node.name.startsWith('getBy') || node.name.startsWith('getAllBy');
+}
+
+// https://github.com/microsoft/TypeScript/issues/16069#issuecomment-369374214
+function isNotNullOrUndefined<T>(input: null | undefined | T): input is T {
+  return input != null;
+}
+
+function mapNodesForWrongGetByQuery(node: TSESTree.CallExpression) {
+  const nodeArguments = node.arguments as ExtendedExpression[];
+  return nodeArguments
+    .map(arg => {
+      if (!arg.callee) {
+        return null;
+      }
+      // Example: `expect(rendered.getBy*)`
+      if (isMemberExpression(arg.callee)) {
+        const node = arg.callee.property as ExtendedExpression['callee'];
+        if (isIdentifier(node) && isUsingWrongQueries(node)) {
+          return node;
+        }
+        return null;
+      }
+
+      // Example: `expect(getBy*)`
+      if (isIdentifier(arg.callee) && isUsingWrongQueries(arg.callee)) {
```

Comment:
This case is represented like this instead:

<img width="1579" alt="image" src="https://user-images.githubusercontent.com/1110551/63999553-096da880-cb04-11e9-869b-c94fb20563fa.png">

Reply:
I could not spot the difference 😅

Reply:
The `arguments` list, one is a `CallExpression` object with a `callee` of type `MemberExpression`, the other one of type `Identifier`.

Reply:
ohhh. now it makes sense. can you make the distinction in the "fixing" part ? in order not to change the first case which would lead to an error.

Reply:
What do you mean?

Reply:
I mean if it is `rendered.getBy..` (so a `MemberExpression `) then fix it, but if it is `getBy` (`CallExpression`) don't fix it. IDK if this works :p

Reply:
Why would you want to do that? Yes, if it's a `CallExpression` you need to change the destructured method as well, but that it not a concern with this rule and it shouldn't be.

It's also stated in the eslint docs:

<img width="1183" alt="image" src="https://user-images.githubusercontent.com/1110551/64002349-1b524a00-cb0a-11e9-8876-5fa05badd8c2.png">

Reply:
makes sense yeah. maybe the second rule should be to not destruct rendered 😄

Reply:
Yeah, that would be a nice rule to have as well. Would you like to give it a try? 😉

Reply:
challenge accepted 😄

Reply:
Do you mean here avoid such `const { getByText } = render();` destructuring?
If yes, I think then we should agree first on why it's considered a bad practice before writing the rule. I personally don't see any harm in doing such destructuring.

Reply:
I have not strongly held opinion about it either. I just thought it'd be fun to write 🤷‍♂ 😅

Reply:
Yes that destructuring.

It's not much about good or bad practice, it's just easier to work with. From my experience by not destructuring you get the following benefits:
- use the variable (e.g. `rendeded`) with autocompletion
- no need to destructure a new function everytime you need one
- makes it easier to swap things out

Reply:
If it's not considered a bad practice, then I wouldn't create a rule prohibiting it. I'm seeing it as a style rule and not a code quality rule. I personally don't have any issues with the points mentioned above.